src/librustc_typeck/outlives/implicit_infer.rs

```diff
@@ -0,0 +1,257 @@
+
+pub fn infer_for_fields<'tcx>(
+    tcx: TyCtxt<'_, 'tcx, 'tcx>,
+    mut inferred_outlives_map: &mut FxHashMap<DefId, Rc<Vec<ty::Predicate<'tcx>>>>,
+) {
+    /*
+        // inferred_outlives_predicate = ['b: 'a] // after round 2
+        struct Foo<'a, 'b> {
+            bar: Bar<'a, 'b> // []
+            // round 2: ['b: 'a] in `required_predicates`
+        }
+
+        // inferred_outlives_predicate = ['b: 'a] // updated after round 1
+        struct Bar<'a, 'b> {
+            field &'a &'b u32 // ['b: 'a] // added to `required_predicates`
+        } // required_predicates = ['b: 'a]
+    */
+
+    let mut changed = true;
+    while changed {
+        changed = false;
+        let mut visitor = InferVisitor {
+            tcx: tcx,
+            inferred_outlives_map: &mut inferred_outlives_map,
+            changed: changed,
+        };
+
+        //iterate over all the crates
+        tcx.hir.krate().visit_all_item_likes(&mut visitor);
+    }
+}
+
+pub struct InferVisitor<'cx, 'tcx: 'cx> {
+    tcx: TyCtxt<'cx, 'tcx, 'tcx>,
+    inferred_outlives_map: &'cx mut FxHashMap<DefId, Rc<Vec<ty::Predicate<'tcx>>>>,
+    changed: bool,
+}
+
+impl<'cx, 'tcx> ItemLikeVisitor<'tcx> for InferVisitor<'cx, 'tcx> {
+    fn visit_item(&mut self, item: &hir::Item) {
+        let def_id = self.tcx.hir.local_def_id(item.id);
+
+        let node_id = self.tcx
+            .hir
+            .as_local_node_id(def_id)
+            .expect("expected local def-id");
+        let item = match self.tcx.hir.get(node_id) {
+            hir::map::NodeItem(item) => item,
+            _ => bug!(),
+        };
+
+        let mut local_required_predicates = FxHashMap();
+        match item.node {
+            hir::ItemUnion(..) |
+            hir::ItemEnum(..) |
+            hir::ItemStruct(..) => {
+                let adt_def = self.tcx.adt_def(def_id);
+                for field_def in adt_def.all_fields() {
+                    let required_predicates: Vec<_> = required_predicates_to_be_wf(
+                        self.tcx,
+                        field_def,
+                        &local_required_predicates,
+                     ).into_iter().collect();
+                    local_required_predicates.insert(
+                        field_def.did,
+                        Rc::new(required_predicates),
+                    );
+                }
+            }
+            _ => {}
+        };
+
+        if local_required_predicates.len() > 0 {
+            self.changed = true;
+            self.inferred_outlives_map
+                .extend(local_required_predicates);
+        }
+    }
+
+    fn visit_trait_item(&mut self, trait_item: &'tcx hir::TraitItem) {}
+
+    fn visit_impl_item(&mut self, impl_item: &'tcx hir::ImplItem) {}
+}
+
+fn required_predicates_to_be_wf<'tcx>(
+    tcx: TyCtxt<'_, 'tcx, 'tcx>,
+    field_def: &ty::FieldDef,
+    infered_outlives_map: &FxHashMap<DefId, Rc<Vec<ty::Predicate<'tcx>>>>,
+) -> HashSet<ty::Predicate<'tcx>> {
```

Comment:
Hmm, so returning a set of *predicates* here is ok, but overly general, since I think we're selecting down to `T: 'a` (`TypeOutlives`) or `'a: 'b` (`RegionOutlives`) predicates. But I guess that's ok..